### PR TITLE
fix: Update current UI instance upon locale change

### LINKED ISSUE #11599: NullPointerException in StateNode when setting locale and using UI.getCurrent() inside localeChange callback.
### Description of the bug / feature
NullPointerException when changing locale
### Minimal reproducible example

Sometimes when changing the locale when Push is enabled, Vaadin throws a NullPointerException. It's difficult to reproduce.

### Expected behavior
No Exception should be thrown
### Actual behavior
A NullPointer Exception is thrown because in StateNode::dumpBeforeClientResponseEntries() there is a check missing for `entries == null`

image

```
java.lang.RuntimeException: Push failed
	at com.vaadin.flow.server.communication.AtmospherePushConnection.push(AtmospherePushConnection.java:200) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.communication.AtmospherePushConnection.push(AtmospherePushConnection.java:175) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.component.UI.push(UI.java:686) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:708) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.VaadinService.requestEnd(VaadinService.java:1474) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1549) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:299) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:109) ~[vaadin-spring-17.0.2.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733) ~[tomcat-embed-core-9.0.45.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.6.jar:5.3.6]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.3.6.jar:5.3.6]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.6.jar:5.3.6]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.3.6.jar:5.3.6]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-
	...
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.ArrayList.isEmpty()" because "entries" is null
	at com.vaadin.flow.internal.StateNode.dumpBeforeClientResponseEntries(StateNode.java:1026) ~[flow-server-7.0.6.jar:7.0.6]
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[na:na]
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1694) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[na:na]
	at com.vaadin.flow.internal.StateTree.flushCallbacks(StateTree.java:407) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.internal.StateTree.runExecutionsBeforeClientResponse(StateTree.java:384) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.communication.UidlWriter.encodeChanges(UidlWriter.java:391) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.communication.UidlWriter.createUidl(UidlWriter.java:174) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.communication.UidlWriter.createUidl(UidlWriter.java:212) ~[flow-server-7.0.6.jar:7.0.6]
	at com.vaadin.flow.server.communication.AtmospherePushConnection.push(AtmospherePushConnection.java:196) ~[flow-server-7.0.6.jar:7.0.6]
	... 46 common frames omitted
```
### Versions:
    - Vaadin / Flow version: 20.0.6
    - Java version: 15
    - OS version: Windows 11
    - Browser version (if applicable): All
    - Application Server (if applicable): Tomcat/Spring
    - IDE (if applicable): IntelliJ

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

Sets the UI on which the locale is being changed as a current UI for the thread. This is needed to have a proper `UI.getCurrent()` instance inside the executed locale change observers in components.

Fixes #11599

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/component/UI.java [hunk 1]
```diff
@@ -1008,6 +1010,11 @@ public ExecutionRegistration beforeClientResponse(Component component,
                     "The 'execution' parameter may not be null");
         }
 
+        if (component.getUI().isPresent() && component.getUI().get() != this) {
```
Comment (Legioth) [inline flow-server/src/main/java/com/vaadin/flow/component/UI.java:1013]: Doesn't seem to have any relationship with "Update current UI instance upon locale change"

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/UI.java:1013]: Well, this can be extracted into another PR

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/UI.java]: Done

File: flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                sameUIResult.setText(Boolean.toString(checkResult));
+                Span caption = new Span(String
+                        .format("Component %d uses current UI instance = ", i));
+                add(caption, sameUIResult, new HtmlComponent(Tag.BR));
+            }
+        }
+
+        private void registerLocaleUpdate(boolean sameUIcheckUp) {
+            VaadinSession session = VaadinSession.getCurrent();
+            session.lock();
```
Comment (denis-anisimov) [inline flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java:110]: Is it needed ?
Every Flow listener should be called under the session lock arelady.

File: flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                localeUpdates.addUICheckResult(sameUIcheckUp);
+                session.setAttribute(LocaleUpdates.class, localeUpdates);
+            } finally {
+                session.unlock();
+            }
+        }
+
+        private LocaleUpdates getLocaleUpdates() {
+            VaadinSession session = VaadinSession.getCurrent();
+            session.lock();
```
Comment (denis-anisimov) [inline flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java:126]: Again: is this needed?

**Review by denis-anisimov — CHANGES_REQUESTED** [review]

File: flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java (hunk 2 above)
Comment (mshabarov) [reply flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java:110]: Yeah, it's not needed. `VaadinSession::setLocale` checks that it already has a lock.

Comment (mshabarov) [reply flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java]: Done

File: flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java (hunk 3 above)
Comment (mshabarov) [reply flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/LocaleChangeView.java]: No.
Done.

**Review by denis-anisimov — APPROVED** [review]

**Pull request merged**